### PR TITLE
Feature: Accessing source code from different sources 

### PR DESCRIPTION
**Pull request by zentox**
This PR provides an interface to access source code form different sources. It includes to subclasses which allows access source code from the submission and from the stdlib.

### CODE REVIEW

File: tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * Returns the path of the source code of the java standard library. The default location is the src.zip file in
+     * the java home directory.
+     *
+     * @return the path of the source code of the java standard library
+     */
+    public static Path getStdlibSource() {
+        try (Stream<Path> stream = Files.walk(Path.of(System.getProperty("java.home"))).parallel()) {
```
[the remark this answers is not in the document]
Comment (zentox) [reply tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java]: Fix in [04faee3](https://github.com/tudalgo/algoutils/pull/88/commits/04faee35dae3168cbcaaae19a457573e9b00bca1)

File: tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/SubmissionJavaResource.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        return resources;
+    }
+
+    private Map<String, String> getContentsByLocal() {
+        // Current test run is a local run
+        Set<Path> excluded = EXCLUDED_DIRECTORIES.stream()
+            .map(source::resolve)
+            .collect(Collectors.toSet());
+
+        try (Stream<Path> paths = Files.walk(source).parallel()) {
```
[the remark this answers is not in the document]
Comment (zentox) [reply tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/SubmissionJavaResource.java]: Fix in [04faee3](https://github.com/tudalgo/algoutils/pull/88/commits/04faee35dae3168cbcaaae19a457573e9b00bca1)

File: tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        if (contents.containsKey(className)) {
+            return contents.get(className);
+        }
+
+        // Case: contents does not contain the class, load it from entries
+        try (ZipFile file = new ZipFile(source.toFile())) {
+            assert cache != null;
+            ZipArchiveEntry entry = cache.get(className);
+            String sourceCode = new String(file.getInputStream(entry).readAllBytes());
```
[the remark this answers is not in the document]
Comment (zentox) [reply tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java:162]: Fix in [04faee3](https://github.com/tudalgo/algoutils/pull/88/commits/04faee35dae3168cbcaaae19a457573e9b00bca1)

File: tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    public Map<String, String> contents() {
+        cache = getCache();
+        if (contents.size() == cache.size()) return contents;
+        // Do not use get(String) since it will create a new ZipFile for each entry
+        try (ZipFile file = new ZipFile(source.toFile())) {
+            cache.entrySet().parallelStream().forEach(entry -> {
+                String className = entry.getKey();
+                ZipArchiveEntry zipEntry = entry.getValue();
+                try {
+                    contents.put(className, new String(file.getInputStream(zipEntry).readAllBytes()));
```
[the remark this answers is not in the document]
Comment (zentox) [reply tutor/src/main/java/org/tudalgo/algoutils/tutor/general/io/StdlibJavaResource.java]: Fix in [04faee3](https://github.com/tudalgo/algoutils/pull/88/commits/04faee35dae3168cbcaaae19a457573e9b00bca1)

**Review by Rdeisenroth — COMMENTED** [review]
Regarding the various zip-actions, i think it might be more effective to use a zip library like zip4j which would imo improve readibility a little.

Also i am a bit confused as to why we would ever need to do source code analysis on the standard library for testing student submissions or load classes outside the source dir. Because having the ability to load arbitrairy classes always raises a big security risk. For example:
- how do we ensure that the loaded source code wasn't tempered? (for example student statically overwriting java.home)
- how can we ensure that no mallicious code get's executed? (assuming the code gets loaded by the class loader)
- potential risk of information leakage (load rubrics from other students or access personal files)
- potential risk of grading manipulation (accessing rivals rubric and giving him 0 points)
- potential risk of replacing own submission files with files of a known to be good student
- potential risk of privilege escalation
- how do we ensure that loading source code doesn't overwrite the existing classes?  

=> So unless we have a specific usecase that requires it i would personally advise against it.

**Comment by Rdeisenroth** [comment]
Currently not Planned.

**Pull request closed without merging**